File: 966-binary-subarrays-with-sum/binary-subarrays-with-sum.cpp
class Solution {
public:
     int numSubarraysWithSum(vector<int>& A, int S) {
        int p=0, i=0, r=0, c=0;
        for(int j = 0; j < A.size(); j++){
            c+= A[j];
            while(c > S && i < j){
                p=0;
                c -= A[i];
                i++;
            }
            while(A[i] == 0 && i < j){
                p++;
                c -= A[i];
                i++;
            }
            if(c == S) r += p+1;
        }
        return r;
    }    
};

File: 966-binary-subarrays-with-sum/binary-subarrays-with-sum.java
class Solution {
    public int numSubarraysWithSum(int[] nums, int goal) {
        return helper(nums,goal)-helper(nums,goal-1);
    }
    int helper(int[] nums,int goal){
        if(goal<0){
            return 0;
        }
        int left=0,right=0,cnt=0,sum=0;
        while(right<nums.length){
            sum+=nums[right];
            while(left<right && sum>goal){
                sum-=nums[left];
                left++;
            }
            if(sum<=goal){
                cnt+=right-left+1;
            }
            right++;
        }
        return cnt;
    }
}